src/OpenCL.cpp

```diff
@@ -148,11 +157,14 @@ __kernel void out_transform(__global float *M, __global net_t *Y,
     int y = 2*block_y;
 
     if (k < K && block < P) {
+        int kHW = k * HW;
         int b = block_y * WTILES + block_x;
+        int bKpad = b*Kpad;
         float temp_m[16];
-        for (int xi = 0; xi < 4; xi++) {
-            for (int nu = 0; nu < 4; nu++) {
-                temp_m[xi*4 + nu] = M[xi*(4*Kpad*Ppad) + nu*(Kpad*Ppad)+ b*Kpad + k];
+        for (int xi = 0; xi < 16; xi+=4) {
+            int xiKPpad = xi * KPpad;
+            for (int nu = 0, nuKPpad = 0; nu < 4; nu++, nuKPpad += KPpad) {
+                temp_m[xi + nu] = M[xiKPpad + nuKPpad + bKpad + k];
```

Comment:
This can be replaced with a single loop.

```
int offset = b*Kpad + k;
for (int xn = 0; xn < 16; xn++) {
    temp_m[xn] = M[xn*KPpad + offset];
}
```

Reply:
But this is worst for the multiplication, then use that same trick like nuKPpad.
```
int offset = b*Kpad + k;
int xnKPpad = 0;
for (int xn = 0, xnKPpad = 0; xn < 16; xn++, xnKPpad += KPpad) {
    temp_m[xn] = M[xnKPpad + offset];
}
```

Reply:
Testing now!

Reply:
I think you could initialize xnKPad with offset to get rid of the addition in array access.

Reply:
Wow, nice job guys!

Reply:
How about
```
for (int xn = 0, xnKPpad = b*Kpad + k; xn < 16; xn++, xnKPpad += KPpad) {
    temp_m[xn] = M[xnKPpad];
}
```